### PR TITLE
Add expense feature

### PR DESCRIPTION
**Pull request by ghost**
1. Add Expense Model (balance/models.py)
2. Add Category Model (balance/models.py)
3. Add urlpatterns in balance/urls.py
4. Add templates in balance/templates/balance
5. Register models in balance/admin.py
6. Add filter expenses feature using django-filter library (balance/filters.py)

### CODE REVIEW

File: balance/forms.py [hunk 1]
```diff
@@ -1,9 +1,15 @@
...
-from . import models
 
+from .models import Expense, ExpenseCategory
 
-class IncomeForm(forms.ModelForm):
+
+class ExpenseForm(forms.ModelForm):
+    class Meta:
+        model = Expense
+        fields = ['expense_name', 'expense_value', 'comment', 'recurring_income']
```
Comment (michal-szastok) [inline balance/forms.py]: change to recurring_expense

File: balance/models.py [hunk 2]
```diff
@@ -1,15 +1,34 @@
...
-    name = models.CharField(max_length=30)
-    value = models.DecimalField(max_digits=6, decimal_places=2)
-    comment = models.CharField(max_length=100)
-    cyclical_income = models.BooleanField()
-    user = models.ForeignKey(User, on_delete=models.CASCADE, related_name='incomes')
+class Expense(models.Model):
+    expense_name = models.CharField(max_length=50)
+    expense_value = models.DecimalField(max_digits=10, decimal_places=2, validators=[MinValueValidator(0.01)])
+    comment = models.TextField(blank=True)
+    cyclical_expense = models.BooleanField()
```
Comment (michal-szastok) [inline balance/models.py]: should be "recurring_expense"

File: balance/models.py [hunk 3]
```diff
@@ -1,15 +1,34 @@
...
+    user = models.ForeignKey(User, on_delete=models.CASCADE, related_name='expenses')
     created_at = models.DateTimeField(auto_now_add=True)
     updated_at = models.DateTimeField(auto_now=True)
+    expense_date = models.DateTimeField(default=datetime.now, blank=True, null=True)
+    expense_category = models.ForeignKey('balance.ExpenseCategory', on_delete=models.CASCADE, related_name="expenses")
```
Comment (michal-szastok) [inline balance/models.py]: on_delete=models.CASCADE
should be
on_delete=models.SET_NULL

File: balance/templates/balance/expense_category_confirm_delete.html [hunk 4]
```diff
@@ -0,0 +1,13 @@
+{% extends 'balance/base.html' %}
+{% block site_title %} Confirm {% endblock %}
+
+{% block content %}
+<form method="post">
+    {% csrf_token %}
+    <p>Are you sure you want to delete "{{ object }}"?</p>
+    {{ form }}
+    <input type="submit" value="Confirm">
+    <button onclick="location.href='{% url 'category_detail' category.id %}'" type="button">Cancel</button>
```
Comment (michal-szastok) [inline balance/templates/balance/expense_category_confirm_delete.html]: 'category_detail'
should be
'expense_category_detail'

File: balance/templates/balance/expense_category_detail.html [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+{% block content %}
+<h2> Category detail</h2>
+
+<p>Category name: {{ object.category_name }}</br>
+    Items in category: NEED TO BE ADDED </br>
+    Total value of category: NEED TO BE ADDED</br></p>
+
+
+<button onclick="location.href='{% url 'category_delete' category.id %}'" type="button">Delete category</button>
```
Comment (michal-szastok) [inline balance/templates/balance/expense_category_detail.html]: 'category_delete'
should be
'expense_category_delete'

File: balance/templates/balance/expense_category_detail.html [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+<button onclick="location.href='{% url 'category_list' %}'" type="button">Back to category list</button>
```
Comment (michal-szastok) [inline balance/templates/balance/expense_category_detail.html]: 'category_list'
should be
'expense_category_list'

File: balance/templates/balance/expense_category_form.html [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+{% block content %}
+<h1>ADD NEW CATEGORY</h1>
+<form method="post" novalidate>
+    {% csrf_token %}
+    {{ form.as_p }}
+    <input type="submit" value="Save">
+</form>
+
+<button onclick="location.href='{% url 'expense_category_list' %}'" type="button">Back to category list</button>
```
Comment (michal-szastok) [inline balance/templates/balance/expense_category_form.html:12]: 'category_list'
should be
'expense_category_list'

File: balance/templates/balance/expense_category_list.html [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+        <td><a href="{% url 'expense_category_detail' category.id %}">{{ category.expense_category_name }}</a></td>
+    </tr>
+    {% empty %}
+    No categories yet
+    {% endfor %}
+    </tbody>
+</table>
+
+<button onclick="location.href='{% url 'expense_category_create' %}'" type="button">Add new category</button>
+<button onclick="location.href='{% url 'expense_list' %}'" type="button">Back to main menu</button>
```
Comment (michal-szastok) [inline balance/templates/balance/expense_category_list.html:25]: 'expense_list'
should be
'expense_category_list'

File: balance/templates/balance/expense_category_list.html [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+{% endblock content %}
```
Comment (michal-szastok) [inline balance/templates/balance/expense_category_list.html]: missing blank line on the end

File: balance/templates/balance/expense_category_form.html [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+
+{% endblock content %}
```
Comment (michal-szastok) [inline balance/templates/balance/expense_category_form.html]: missing blank line on the end

File: balance/templates/balance/expense_list.html [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+        category
+    </button>
+</div></p>
+
+<p>
+<div class="main-menu-button text-center">
+    <button onclick="location.href='#'" type="button" class="btn btn-primary">Back to main menu</button>
+</div></p>
+
+{% endblock content %}
```
Comment (michal-szastok) [inline balance/templates/balance/expense_list.html]: missing blank line

File: balance/views.py [hunk 12]
```diff
@@ -1,18 +1,97 @@
...
+
+class ExpenseUpdateView(UpdateView):
+    model = Expense
+    fields = '__all__'
+    success_url = reverse_lazy('expense_list')
+
+
+class ExpenseDeleteView(DeleteView):
+    model = Expense
+    success_url = reverse_lazy('expense_list')
```
Comment (michal-szastok) [inline balance/views.py:66]: add:
    def get_queryset(self):
        qs = super(IncomeCategoryDeleteView, self).get_queryset()
        return qs.filter(user=self.request.user)

File: balance/views.py [hunk 13]
```diff
@@ -1,18 +1,97 @@
...
+    def form_valid(self, form):
+        form.instance.user = self.request.user
+        return super().form_valid(form)
+
+
+class ExpenseCategoryDeleteView(DeleteView):
+    model = ExpenseCategory
+    fields = '__all__'
+    template_name = 'balance/expense_category_confirm_delete.html'
+    success_url = reverse_lazy('expense_category_list')
```
Comment (michal-szastok) [inline balance/views.py:114]: add:
    def get_queryset(self):
        qs = super(IncomeCategoryDeleteView, self).get_queryset()
        return qs.filter(user=self.request.user)

**Review by michal-szastok — CHANGES_REQUESTED** [review]
Need small improvement

**Comment by ghost** [comment]
Thanks for reviewing code, I've made needed changes.

**Pull request merged**